src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,613 @@
+\\begin{aligned}
+    \\langle U | \\hat{H}_1| U \\rangle & = 
+    \\sum_{i=1}^{N} h_{ii} \\langle \\psi_i | a^{\\dagger}_{i} a_{i} | \\psi_j \\rangle  \\\\
+    & = \\sum_{i=1}^{n_{\\alpha}} h_{ii} 
+        \\langle \\varphi_i | a^{\\dagger}_{i} a_{i} | \\varphi_i \\rangle 
+      + \\sum_{i=1}^{n_{\\beta}} h_{ii} 
+        \\langle \\varphi_i | a^{\\dagger}_{i} a_{i} | \\varphi_i \\rangle.
+\\end{aligned}
+```
+"""
+function one_electron_integral(
+    int1::Array{T,2},
+    occ_modes::Tuple{AbstractVector{FermiFSIndex},AbstractVector{FermiFSIndex}}
+)::T where {T<:Number}
+    one_elec_int = zero(T)
+    for occ_mode in occ_modes
+        for i in occ_mode
+            one_elec_int += int1[i.mode, i.mode]
+        end
+    end
+    one_elec_int
+end
+
+"""
+    two_electron_integral(
+        int2::Array{T,4},
+        occ_modes::Tuple{AbstractVector{FermiFSIndex},AbstractVector{FermiFSIndex}}
+    )::T where {T<:Number}
+
+Calculate the two body operator diagonal term ``\\langle U | \\hat{H}_2| U \\rangle`` 
+with Slater-Condon rules.
+
+```math
+\\begin{aligned}
+    \\langle U | \\hat{H}_2| U \\rangle 
+        & = \\sum_{i<j}^{n_\\alpha} [\\langle ij|ij\\rangle - \\langle ij|ji\\rangle] \\\\
+        & + \\sum_{i<j}^{n_\\beta} [\\langle ij|ij\\rangle - \\langle ij|ji\\rangle] \\\\
+        & + \\sum_{i}^{n_\\alpha}\\sum_{j}^{n_\\beta} \\langle ij|ij\\rangle. \\\\
+\\end{aligned}
+```
+"""
+function two_electron_integral(
+    int2::Array{T,4},
+    occ_modes::Tuple{AbstractVector{FermiFSIndex},AbstractVector{FermiFSIndex}}
+)::T where {T<:Number}
+    two_elec_int = zero(T)
+
+    sum_alpha_alpha = zero(T)
+    for i in occ_modes[1]
+        for j in occ_modes[1]
+            if i.mode ≠ j.mode
+                sum_alpha_alpha += int2[i.mode, j.mode, i.mode, j.mode] -
+                                   int2[i.mode, j.mode, j.mode, i.mode]
+            end
+        end
+    end
+    two_elec_int += 0.5 * sum_alpha_alpha
+
+    sum_alpha_beta = zero(T)
+    for i in occ_modes[1]
+        for j in occ_modes[2]
+            sum_alpha_beta += int2[i.mode, j.mode, i.mode, j.mode]
+        end
+    end
+    two_elec_int += sum_alpha_beta
+
+    sum_beta_beta = zero(T)
+    for i in occ_modes[2]
+        for j in occ_modes[2]
+            if i.mode ≠ j.mode
+                sum_beta_beta += int2[i.mode, j.mode, i.mode, j.mode] -
+                                 int2[i.mode, j.mode, j.mode, i.mode]
+            end
+        end
+    end
+    two_elec_int += 0.5 * sum_beta_beta
+
+    two_elec_int
+end
+
+@inline function flip_spin_chan(chan::Int)::Int
+    if chan == 1
+        return 2
+    elseif chan == 2
+        return 1
+    end
+end
+
+"""
+This struct is used internally to represent the iterator for 
+off-diagonal terms generation.
+"""
+struct MolecularHamiltonianOffDiagonalsIterator{
+    T,A<:FermiFS2C,D,OA,OB,UA,UB,TI<:FermiFSIndex
+}
+    address::A
+    op::MolecularHamiltonian{T,A,D}
+    modes::Modes{OA,OB,UA,UB,TI}
+end
```

Comment:
A suggestion to rename (simplify the name) and declutter the type parameters:
```suggestion
struct MolecularHamiltonianOffDiagonals{
    T,A<:FermiFS2C,H<:MolecularHamiltonian{T,A},M<:Modes
}
    address::A
    op::H
    modes::M
end
```